### PR TITLE
[MPS] Fuse cross_entropy into a native Metal kernel

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
# [MPS] Fuse cross_entropy into a native Metal kernel

> **Reviewing just this PR.** Stacked leaf (cumulative-leaf fork on `main`), so the GitHub diff
> shows the whole softmax stack **+** this one. To see **only what this PR adds on top of
> #187457**, use this fork compare (renders as a normal diff of just this delta):
> **https://github.com/anagnorisis2peripeteia/pytorch/compare/mps-logsoftmax-pr2...mps-crossentropy-pr3**
>
> This PR's own delta — **11 files, +1002**: `CrossEntropyKernel.metal` (+348), `CrossEntropy.mm`
> (+199), `native_functions.yaml` (op reg), `torch/_inductor/decomposition.py` (+106),
> `derivatives.yaml`, `test_mps.py`, `HasDecompTest.expect`, `common_modules.py`, bench.

## What / why

Adds a fused 2D cross-entropy Metal kernel that computes the per-sample loss with a single
online log-sum-exp pass over the `[B, V]` logits, **never materializing the full `[B, V]`
log_softmax intermediate** (the decomposed `log_softmax + nll_loss` path allocates and writes
it twice). Backward recomputes softmax from the saved per-row LSE. New ops
`_fused_cross_entropy_loss_2d_forward` / `_backward` (MPS + Meta impls), plus an inductor
decomposition so `torch.compile` lowers `F.cross_entropy` to the fused op. A
`PYTORCH_MPS_FORCE_DECOMPOSED_CROSS_ENTROPY` escape hatch falls back to the reference path.
Stacked on log_softmax.

Supports `weight`, `ignore_index`, and `label_smoothing`. The inductor decomposition casts a
non-long target's gather index to `long` (mirroring eager `target.to(kLong)`), so a `uint8`
target under `torch.compile` matches eager.

## Performance

Benchmark: `benchmarks/mps/bench_crossentropy.py` (committed). Same-build A/B, mode A =
decomposed `log_softmax + nll_loss` (forced via `PYTORCH_MPS_FORCE_DECOMPOSED_CROSS_ENTROPY=1`),
mode B = fused Metal; 3 reps, median per cell, amortized many-iter/one-sync. M4 Pro. 60 cells
(batch x vocab x dtype x contig/noncontig x fwd / fwd+bwd). **57 win, 3 below 0.95x.** The
wins concentrate on realistic training shapes (large vocab, fwd+bwd):

| cell (batch x vocab) | decomposed | fused | speedup |
|---|---|---|---|
| 16384x50304 f16 fwd+bwd | 101,809 us | 21,353 us | 4.77x |
| 16384x50304 bf16 fwd+bwd | 99,530 us | 21,324 us | 4.67x |
| 2048x128256 f16 fwd+bwd | 29,698 us | 6,919 us | 4.29x |
| 4096x32000 f16 fwd+bwd | 14,347 us | 3,507 us | 4.09x |
| 4096x1024 f16 fwd (small vocab) | 80 us | 119 us | 0.67x |
| 16384x50304 f32 noncontig fwd+bwd | 818,074 us | 862,335 us | 0.95x |

The 3 sub-0.95x cells are small-vocab (V=1024) forward-only (the fused kernel's fixed
overhead is not amortized when there is little to fuse; forward-only + tiny vocab is
uncommon in training) and one marginal non-contiguous f32 case. Both are covered by the
decomposed escape hatch.

## Correctness / Test Plan

- `PYTORCH_TEST_MPS=1 python -m pytest test/test_mps.py -k "cross_entropy_metal" -q`
- `test_cross_entropy_metal_paths`: uint8 target under `torch.compile` (decomp gather-index
  cast), C>256 multi-simdgroup forward, and label smoothing fwd+bwd vs fp32 CPU.
- Full `--mps` suite (incl. inductor) -- clean modulo the unrelated
  `test_metal.py::...test_conv`.

## BC-breaking?

No public API change. `F.cross_entropy` numerics are unchanged (fused matches the decomposed
reference); the fused op is an internal implementation detail selected on MPS.

## Review history

Local review + gate dispositions (autoreview + local_clawsweeper, including the `ignore_index`
int32 truncation fix): https://gist.github.com/9d1528bdd58623d9d5653024ae2a4024

---

## Update: double-backward, symbolic ignore_index, and LSE NaN guards

Responding to review, three fixes:

- **Double backward restored**: `_fused_cross_entropy_loss_2d_backward` now has an analytic derivative (FunctionsManual) — the softmax-Jacobian total derivative through the saved `lse`, following the `native_layer_norm_backward` convention, with the `grad_loss` leg contracted via the backward-at-ones pattern. Validated to ~1e-16 against autograd in fp64; new test covers both mask branches CPU-vs-MPS plus a triple-backward smoke.
- **Export/dynamic shapes**: the eligibility gate previously concretized a symbolic `ignore_index` before checking `is_mps`, specializing dynamic batch dims on non-MPS/export paths. `guard_int` now happens inside the gate after the device/shape rejections; `torch.export` keeps the batch Dim dynamic and `torch.compile(dynamic=True)` records a single frame.
- **LSE NaN guards**: the fused kernel's per-thread online recurrence produced `0 * exp(NaN)` for all-(-inf) prefix chunks (left-padded masked rows) and dropped NaN logits via `fmax`; both fixed with the same guard+flag pattern as the softmax stack (parity repros for both cases).

Review log: https://gist.github.com/anagnorisis2peripeteia/75aa5844c4d9d74885bbc9d12a58c2f6

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related to this PR, and review the changes

**Comment by anagnorisis2peripeteia** [comment]
Fixed in 1fb3e48.

- **Non-MPS link break:** moved both `_fused_cross_entropy_loss_2d_{forward,backward}_meta` out of the MPS-only `CrossEntropy.mm` into the device-agnostic `LossNLL.cpp` (with the two `_native.h` includes for the prototypes), so the `Meta` dispatch — registered on every build for shape tracing — links on linux/cuda/rocm/xpu. Mirrors `copy_meta` (Copy.cpp) / `ctc_loss_meta` (LossCTC.cpp); MPS forward/backward stay in the `.mm`. Verified it builds and the cross_entropy / softmax MPS tests pass.
- **pyrefly (`unbound-name`):** on the label-smoothing path `w` was only bound in the first `if weight is not None` block; the second one now uses `weight.float()` directly inside the narrowed guard. `pyrefly check` is clean.

`pr-sanity-checks` is the cumulative-diff-size gate — this PR shows softmax + log_softmax + CE because it's stacked, and it resolves as the parents land.

Thanks for the catch — the cross-platform link issue isn't visible from a Mac-only build.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by anagnorisis2peripeteia** [comment]
@jansel thanks for taking a look. Confirmed: the only failure caused by this PR was `test_decomp.py::HasDecompTest::test_has_decomposition`. The two new ops (`_fused_cross_entropy_loss_2d_forward`/`_backward`) have a `Meta` kernel but their decomposition is registered in the inductor table (`torch/_inductor/decomposition.py`), not core `torch._decomp`, so they need to be listed in `test/expect/HasDecompTest.test_has_decomposition.expect`. Added both (in sorted order) in `ae2a19bb1ef` and verified locally:

```
$ python test/test_decomp.py HasDecompTest.test_has_decomposition
.  (1 passed)
```

The rest of the red is unrelated to this PR: broken-trunk + infra (container init, the Clang-14 runner vs required Clang-16+), the `pr-sanity-checks` stacked-diff size gate, and the Dr.CI-flagged flaky jobs. Rebasing onto current `main` clears the broken-trunk set.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by anagnorisis2peripeteia** [comment]
Moved back to draft: rebasing the stack onto current main (base was stale, causing the DIRTY state) and re-running the full gate suite before re-opening for review.

**Comment by anagnorisis2peripeteia** [comment]
@jansel you know this is part of a stack right?

**Comment by anagnorisis2peripeteia** [comment]
Closing this PR as a dead stack, not as giving up on fused cross-entropy.

GitHub is showing the softmax core, softmax perf, **log_softmax (#187457)**, then CE. #187457 is closed: Isalia20 #195441 owns log_softmax on current `main`. Reviewing this PR would re-introduce our log_softmax kernel on top of that.

Fused CE is still needed (`SoftMax.mm` / `Activation.mm` log_softmax + CE graph paths remain until those land). Rebuild CE after #187456 (softmax core) and #195441 (their log_softmax), as a CE-only diff. Tracked in #187455.
